src/build.d

```diff
@@ -1447,23 +1386,38 @@ class DependencyRef
     }
 }
 
+/** Initializes an object using a chain of method calls */
+struct MethodInitializer(T)
+{
+    private T obj;
+    auto ref opDispatch(string name, T)(T arg)
+    {
+        mixin("obj." ~ name ~ " = arg;");
+        return this;
+    }
+}
+
 /**
 Takes a lambda and returns a memoized function to build a dependecy object.
+The lambda takes a builder and a dependency object.
+This differs from makeDepWithArgs in that the function literal does not need explicit
+parameter types.
 */
-alias makeDep(alias Func) = memoize!(function() {
-    static struct DepBuilder
-    {
-        private DependencyRef dep;
-        auto ref opDispatch(string name, T)(T arg)
-        {
-            mixin("dep." ~ name ~ " = arg;");
-            return this;
-        }
-    }
-    auto builder = DepBuilder(new DependencyRef());
-    Func(&builder, builder.dep);
-    builder.dep.finalize();
-    return builder.dep;
+alias makeDep(alias Func) = makeDepImpl!(Func);
+
+/**
+Takes a lambda and returns a memoized function to build a dependecy object.
+The lambda takes a builder, dependency object and any extra arguments needed
+to create the dependnecy.
+This differs from makeDep in that the function literal must contain explicit parameter types.
+*/
+alias makeDepWithArgs(alias Func) = makeDepImpl!(Func, Parameters!Func[2..$]);
+
+alias makeDepImpl(alias Func, T...) = memoize!(function(T args) {
+    auto builder = MethodInitializer!DependencyRef(new DependencyRef());
+    Func(&builder, builder.obj, args);
```

Comment:
```suggestion
    Func(builder, builder.obj, args);
```

Reply:
Good suggestion.  Will include in the next PR.